File: Sources/HedosKernel/Discovery/GGUFShards.swift
import Foundation

enum GGUFShards {
    struct Member: Hashable {
        let index: Int
        let url: URL
        let bytes: Int64
    }

    struct ShardGroup {
        let base: String
        let total: Int
        let members: [Member]

        var firstShard: URL? { members.first(where: { $0.index == 1 })?.url }
        var footprintBytes: Int64 { members.reduce(0) { $0 + $1.bytes } }
        var complete: Bool { members.count == total }
    }

    static func parse(_ filename: String) -> (base: String, index: Int, total: Int)? {
        guard filename.lowercased().hasSuffix(".gguf") else { return nil }
        let stem = String(filename.dropLast(".gguf".count))
        guard let ofRange = stem.range(of: "-of-", options: .backwards) else { return nil }
        let totalField = String(stem[ofRange.upperBound...])
        guard totalField.count == 5, let total = Int(totalField), total > 0 else { return nil }
        let head = stem[..<ofRange.lowerBound]
        guard let dashRange = head.range(of: "-", options: .backwards) else { return nil }
        let indexField = String(head[dashRange.upperBound...])
        guard indexField.count == 5, let index = Int(indexField), index > 0, index <= total
        else { return nil }
        let base = String(head[..<dashRange.lowerBound])
        guard !base.isEmpty else { return nil }
        return (base, index, total)
    }

    static func name(base: String, index: Int, total: Int) -> String {
        String(format: "%@-%05d-of-%05d.gguf", base, index, total)
    }

    static func group(_ files: [(url: URL, bytes: Int64)]) -> (
        groups: [ShardGroup], loose: [URL]
    ) {
        struct Key: Hashable {
            let directory: String
            let base: String
            let total: Int
        }

        var buckets: [Key: [Member]] = [:]
        var loose: [URL] = []
        for file in files {
            guard let shard = parse(file.url.lastPathComponent) else {
                loose.append(file.url)
                continue
            }
            let key = Key(
                directory: file.url.deletingLastPathComponent().path,
                base: shard.base, total: shard.total)
            buckets[key, default: []].append(
                Member(index: shard.index, url: file.url, bytes: file.bytes))
        }

        let groups = buckets.map { key, members in
            ShardGroup(
                base: key.base, total: key.total,
                members: members.sorted { $0.index < $1.index })
        }
        return (groups, loose)
    }

    static func discoveredModels(
        from files: [(url: URL, bytes: Int64)], kind: SourceKind, repo: (URL) -> String?
    ) -> (discovered: [DiscoveredModel], issues: [String]) {
        let (groups, loose) = group(files)
        let bytesByURL = Dictionary(files.map { ($0.url, $0.bytes) }, uniquingKeysWith: { a, _ in a })
        let hint = ModalityHints.gguf
        var discovered: [DiscoveredModel] = []
        var issues: [String] = []

        for url in loose {
            discovered.append(
                DiscoveredModel(
                    name: url.deletingPathExtension().lastPathComponent,
                    source: ModelSource(kind: kind, path: url.path, repo: repo(url)),
                    modalityHint: hint.modality,
                    capabilitiesHint: hint.capabilities,
                    executionHint: hint.execution,
                    footprintBytes: bytesByURL[url] ?? 0,
                    primaryWeightPath: url.path))
        }
        for shardGroup in groups {
            guard let first = shardGroup.firstShard else {
                issues.append(
                    "sharded model \(shardGroup.base) is missing its first part — skipped")
                continue
            }
            discovered.append(
                DiscoveredModel(
                    name: shardGroup.base,
                    source: ModelSource(kind: kind, path: first.path, repo: repo(first)),
                    modalityHint: hint.modality,
                    capabilitiesHint: hint.capabilities,
                    executionHint: hint.execution,
                    footprintBytes: shardGroup.footprintBytes,
                    primaryWeightPath: first.path,
                    downloading: !shardGroup.complete))
        }
        return (discovered, issues)
    }
}

File: kernel/src/discovery/gguf_shards.rs
//! Recognizing and grouping multi-part GGUF weight files named like
//! `model-00001-of-00005.gguf`.

use std::collections::{BTreeMap, BTreeSet};
use std::path::{Path, PathBuf};

/// The parts of a sharded GGUF filename.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ShardName {
    /// The base name shared by every shard in the set.
    pub base: String,
    /// This shard's 1-based index.
    pub index: usize,
    /// The total number of shards declared in the name.
    pub total: usize,
}

/// One member file of a shard group.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Member {
    /// The shard's 1-based index.
    pub index: usize,
    /// The file's path.
    pub path: PathBuf,
    /// The file's size in bytes.
    pub bytes: i64,
}

/// A set of shards that together form one model.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ShardGroup {
    /// The shared base name.
    pub base: String,
    /// The declared total shard count.
    pub total: usize,
    /// The member files present, sorted by index.
    pub members: Vec<Member>,
}

impl ShardGroup {
    /// The path of the first shard (index 1), if present.
    pub fn first_shard(&self) -> Option<&Path> {
        self.members
            .iter()
            .find(|member| member.index == 1)
            .map(|member| member.path.as_path())
    }

    /// The total size of all present members.
    pub fn footprint_bytes(&self) -> i64 {
        self.members.iter().map(|member| member.bytes).sum()
    }

    /// Whether every declared shard is present.
    pub fn complete(&self) -> bool {
        self.members.len() == self.total
    }
}

/// Parse a `<base>-<index>-of-<total>.gguf` filename. The index and total fields
/// are exactly five digits, `1 <= index <= total`, and the base is non-empty.
pub fn parse(filename: &str) -> Option<ShardName> {
    let cut = filename.len().checked_sub(".gguf".len())?;
    if !filename.get(cut..)?.eq_ignore_ascii_case(".gguf") {
        return None;
    }
    let stem = &filename[..cut];
    let of_position = stem.rfind("-of-")?;
    let total_field = &stem[of_position + "-of-".len()..];
    let total = five_digit_number(total_field)?;
    if total == 0 {
        return None;
    }
    let head = &stem[..of_position];
    let dash_position = head.rfind('-')?;
    let index_field = &head[dash_position + 1..];
    let index = five_digit_number(index_field)?;
    if index == 0 || index > total {
        return None;
    }
    let base = &head[..dash_position];
    if base.is_empty() {
        return None;
    }
    Some(ShardName {
        base: base.to_owned(),
        index,
        total,
    })
}

/// Build the canonical filename for a shard.
pub fn shard_filename(base: &str, index: usize, total: usize) -> String {
    format!("{base}-{index:05}-of-{total:05}.gguf")
}

/// Split files into shard groups and loose (non-sharded) files. Shards are
/// grouped by directory, base name, and declared total, and each group's members
/// are sorted by index.
pub fn group(files: &[(PathBuf, i64)]) -> (Vec<ShardGroup>, Vec<PathBuf>) {
    let mut buckets: BTreeMap<(String, String, usize), Vec<Member>> = BTreeMap::new();
    let mut loose: Vec<PathBuf> = Vec::new();

    for (path, bytes) in files {
        let shard = path
            .file_name()
            .and_then(|name| name.to_str())
            .and_then(parse);
        match shard {
            Some(shard) => {
                let directory = path
                    .parent()
                    .map(|dir| dir.to_string_lossy().into_owned())
                    .unwrap_or_default();
                buckets
                    .entry((directory, shard.base, shard.total))
                    .or_default()
                    .push(Member {
                        index: shard.index,
                        path: path.clone(),
                        bytes: *bytes,
                    });
            }
            None => loose.push(path.clone()),
        }
    }

    let groups = buckets
        .into_iter()
        .map(|((_, base, total), members)| {
            let mut seen = BTreeSet::new();
            let mut members: Vec<Member> = members
                .into_iter()
                .filter(|member| seen.insert(member.index))
                .collect();
            members.sort_by_key(|member| member.index);
            ShardGroup {
                base,
                total,
                members,
            }
        })
        .collect();
    (groups, loose)
}

fn five_digit_number(field: &str) -> Option<usize> {
    if field.len() != 5 || !field.bytes().all(|byte| byte.is_ascii_digit()) {
        return None;
    }
    field.parse().ok()
}
